src/compiler/types.ts

```diff
@@ -6908,6 +6908,9 @@ export interface Diagnostic extends DiagnosticRelatedInformation {
     /** @internal */ skippedOn?: keyof CompilerOptions;
 }
 
+/** @internal */
+export type DiagnosticArguments = (string | number | undefined)[];

```

Comment:
Allowing `undefined` feels too permissive to be honest.

Reply:
You're totally right, but, the bottommost calls already accepted `undefined`, and if I don't keep it, it's not possible to write a function which takes in a message and an optional parameter. To make that work, we'd need to strictly type our diagnostics to encode how many arguments they take, which technically sounds good but I'm guessing will be annoying.

Reply:
But, let me try and make that change again and see how bad it is.

Reply:
Well, I gave it a shot, but I'd probably want to split some of it into another PR.

Reply:
FYI I did this; it was slightly more invasive than I hoped because `tokenToString` may return `undefined`, so I had to give it a stronger overload and fix up other code to make sure that we do it right.